File: 0078-subsets/0078-subsets.java
class Solution {
    public void solve(int[] nums, int i, List<Integer> curr, List<List<Integer>> ans){
        if(i == nums.length){
            ans.add(new ArrayList<>(curr));
            return ;
        }

        curr.add(nums[i]);
        solve(nums, i+1, curr, ans);
        curr.remove(curr.size()-1);

        solve(nums, i+1, curr, ans);
        
    }
    public List<List<Integer>> subsets(int[] nums) {
        List<List<Integer>> ans = new ArrayList<>();
        List<Integer> curr = new ArrayList<>();
        solve(nums, 0, curr, ans);

        return ans;
    }
}

File: 0078-subsets/0078-subsets.cpp
class Solution {
public:
    void solve(vector<int>& nums, int i , vector<int>& curr, vector<vector<int>> &res){
        if(i == nums.size()){
            res.push_back(curr);
            return ;
        }

        // not take
        solve(nums, i+1, curr, res);

         // take
         curr.push_back(nums[i]);
         solve(nums, i+1, curr, res);
         curr.pop_back();
    }
    vector<vector<int>> subsets(vector<int>& nums) {
        vector<vector<int>> res;
        vector<int> curr;
        solve(nums, 0, curr, res);
        return res;
    }
};